allennlp/sanity_checks/batch_norm_verification.py

```diff
@@ -0,0 +1,85 @@
+"""
+Code based almost entirely on
+https://github.com/awaelchli/pytorch-lightning-snippets/commit/7db53f774715d635c59ef56f21a17634d246b2c5
+"""
+
+import torch
+from torch import nn as nn
+from typing import Tuple, List, Callable
+from allennlp.sanity_checks.verification_base import VerificationBase
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class BatchNormVerification(VerificationBase):
+
+    normalization_layer = (
+        nn.BatchNorm1d,
+        nn.BatchNorm2d,
+        nn.BatchNorm3d,
+        nn.SyncBatchNorm,
+        nn.InstanceNorm1d,
+        nn.InstanceNorm2d,
+        nn.InstanceNorm3d,
+        nn.GroupNorm,
+        nn.LayerNorm,
+    )
+
+    def __init__(self, model: nn.Module):
+        super().__init__(model)
+        self._hook_handles: List[Callable] = []
+        self._module_sequence: List[Tuple[str, nn.Module]] = []
+        self._detected_pairs: List[Tuple] = []
+
+    @property
+    def detected_pairs(self) -> List[Tuple]:
+        return self._detected_pairs
+
+    def check(self, inputs) -> bool:  # type: ignore
+        inputs = self._get_inputs_copy(inputs)
+        self.register_hooks()
+        # trigger the hooks and collect sequence of layers
+        self._model_forward(inputs)
+        self.destroy_hooks()
+        self.collect_detections()
+        return not self._detected_pairs
+
+    def collect_detections(self):
+        detected_pairs = []
+        for (name0, mod0), (name1, mod1) in zip(
+            self._module_sequence[:-1], self._module_sequence[1:]
+        ):
+            bias = getattr(mod0, "bias", None)
+            detected = (
+                isinstance(mod1, self.normalization_layer)
+                and mod1.training  # TODO: do we want/need this check?
```

Comment:
I think the whole thing only makes sense if we're training. When evaluating, you might want to run all kinds of crazy stuff, because you're making baselines of somebody else's code.